ValveResourceFormat/Serialization/VfxEval.cs

```diff
@@ -0,0 +1,503 @@
+                Expressions.Push($"exists({extVarname})");
+                return;
+            }
+
+            // parser terminates here
+            if (op == OPCODE.ENDOFDATA)
+            {
+                if (dataReader.PeekChar() != -1)
+                {
+                    ErrorMessage = "malformed data!";
+                    ErrorWhileParsing = true;
+                    return;
+                }
+                var finalExp = Expressions.Pop();
+                while (finalExp.Length > 2 && finalExp[0] == '(' && finalExp[finalExp.Length - 1] == ')')
+                {
+                    finalExp = Trimb(finalExp);
+                }
+                DynamicExpressionList.Add($"return {finalExp};");
+                return;
+            }
+
+            // this point should never be reached
+            // throw new Exception($"UNKNOWN OPCODE = 0x{(int)op:x2}, offset = {dataReader.BaseStream.Position}");
+            ErrorWhileParsing = true;
+            ErrorMessage = $"UNKNOWN OPCODE = 0x{(int)op:x2}, offset = {dataReader.BaseStream.Position}";
+        }
+
+        private void ApplyFunction(string funcName, int nrArguments)
+        {
+            if (nrArguments == 0)
+            {
+                Expressions.Push($"{funcName}()");
+                return;
+            }
+            string exp1 = Expressions.Pop();
+            if (nrArguments == 1)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp1)})");
+                return;
+            }
+            string exp2 = Expressions.Pop();
+            if (nrArguments == 2)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp2)},{Trimb(exp1)})");
+                return;
+            }
+            string exp3 = Expressions.Pop();
+            if (nrArguments == 3)
+            {
+                // trim or not to trim ...
+                Expressions.Push($"{funcName}({Trimb(exp3)},{Trimb(exp2)},{Trimb(exp1)})");
+                // expressions.Push($"{funcName}({exp3},{exp2},{exp1})");
+                return;
+            }
+            string exp4 = Expressions.Pop();
+            if (nrArguments == 4)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp4)},{Trimb(exp3)},{Trimb(exp2)},{Trimb(exp1)})");
+                return;
+            }
+
+            throw new Exception("this cannot happen!");
+        }
+
+        private static string GetSwizzle(byte b)
+        {
+            string[] axes = { "x", "y", "z", "w" };
+            var swizzle = axes[b & 3] + axes[(b >> 2) & 3] + axes[(b >> 4) & 3] + axes[(b >> 6) & 3];
+            var i = 3;
+            while (i > 0 && swizzle[i - 1] == swizzle[i])
+            {
+                i--;
+            }
+            return swizzle.Substring(0, i + 1);
+        }
+
+        private static string Trimb(string exp)
+        {
+            return exp[0] == '(' && exp[^1] == ')' ? exp[1..^1] : exp;
+        }
+        //private string Trimb2(string exp)
```

Comment:
Remove this

Reply:
you mean the comment .. right. I'll take it out